rfcs/007-bytecode-interpreter.md

```diff
@@ -0,0 +1,839 @@
+optimized for currying and partial application of closures: it uses a
+"push-enter" evaluation strategy, as in the Krivine machine, where the
+traditional call-by-value machines such as the SECD uses "eval-apply".
+
+Arguments and local variables are stored on the stack (the volatile part of the
+environment), and other captured variables are saved on the heap together with
+closures.
+
+The ZAM has 145 instructions where many of them are variations of the same
+operation (`APPLY`, `APPLY1`, `APPLY2` and `APPLY3` for instance). They are
+reduced to rather low-level operation: environment manipulation, function
+application, boolean and integers operations, branching, exceptions, memory
+allocation/value creation.
+
+While the ZAM is considered stack-based, it has a few predefined registers:
+
+- PC: program counter
+- SP: stack pointer
+- ACCU: accumulator
+- TRAPSP: stack pointer of highest exception handler
+- EXTRA_ARGS: number of extra arguments to function application
+- ENV: environment
+- GLOBAL: global data
+
+#### References
+
+- [Real World OCaml chapter on bytecode](https://dev.realworldocaml.org/compiler-backend.html),
+- [Original Leroy's paper: _The ZINC experiment: an economical implementation of
+  the ML language_ (1990)](https://inria.hal.science/inria-00070049/document)
+- [OCaml bytecode instructions](http://cadmium.x9c.fr/distrib/caml-instructions.pdf)
+- [OCaml memory representation of values](https://ocaml.org/docs/memory-representation)
+- [From Krivine's machine to the Caml implementations](https://xavierleroy.org/talks/zam-kazam05.pdf)
+- [OCaml's representation of closures](https://github.com/Gbury/ocaml-memgraph/blob/master/doc/closures.md)
+
+### Haskell
+
+Despite not being advertised, Haskell has an interpreter as well, which is used
+mostly for the GHCi REPL. This section describes what we know of the actual the
+bytecode interpreter, but also incorporates some aspects of the original STG
+machine which is the basis of the low-level operational semantics of Haskell.
+
+#### Memory representation
+
+Closures (thunks and functions) are represented uniformly as an environment and
+a pointer to a closure info table. The closure info table contains the code
+pointer and other garbage-collection related data. The closure info table is
+generated only once per binding appearing in the code and thus shared by dynamic
+instances of the same code. In some other machines for lazy lambda calculi,
+thunks hold additional data such as a flag (evaluated or suspended) and a
+potential value that is filled after the thunk has been forced. When accessing a
+thunk, the caller checks the state, and then either retrieve the value or
+initiate an update sequence and retrieve the code of the unevaluated expression.
+In Haskell, the update process is performed _by the callee_ instead (the thunk's
+code), such that thunk access is uniform: it's an unconditional jump to the
+corresponding code.
```

Comment:
In fact, when pointing to an info table, you actually point directly to the code pointer. The metadata in the info table is accessed backwards by _subtracting_ from the pointer. This way entering a thunk is an indirect jump, which many processors support in a single instruction.

Reply:
Ah, right; I think it's mentioned in the STG paper that this was difficult to do with the approach proposed there going through ANSI C, but easy to do with a custom native code generation backend, which I suppose is what GHC does today.